### PR TITLE
[python] make `early_stopping` callback pickleable

### PR DESCRIPTION
**Pull request by Yard1**
Hey, I'm a maintainer of [LightGBM-Ray](https://github.com/ray-project/lightgbm_ray). I was working on early stopping support for that library, however the early stopping callback on LightGBM's master branch raises exceptions in [our tests (see `test_linux_cutting_edge (3.7)`)](https://github.com/ray-project/lightgbm_ray/runs/5222443116?check_suite_focus=true#logs).

I assumed this is due to the `nonlocal` variables present in the callback and them not interacting well with Ray serialization using cloudpickle. I would imagine there would be similar issues with other distributed computing or serialization libraries.

In order to solve that issue, I took the liberty of rewriting that callback into a callable class, with the `nonlocal` variables becoming attributes. This solved the issues with serialization. No functionality has been affected.

Please let me know if this approach is suitable, and if not, I would love to discuss other potential solutions to the problem I am facing.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
@Yard1 Hey, thank you very much for the interest in LightGBM and this contribution!

As a early opinion I would say I support conversion complex functions with closures into classes. I think this will make callbacks usage more user friendly and LightGBM codebase simpler.

Just a note: we have received a feedback about the need in callbacks serializability from Neptune maintainers recently #4719. Also, XGBoost migrated to callback classes: https://github.com/dmlc/xgboost/pull/6199. So I believe this is the right direction.

**Comment by jameslamb** [comment]
I agree with @StrikerRUS , I definitely support the proposed direction. I'll mark this PR as "in-progress" for now. Please get the unit tests working again, and then we'll be able to provide a review. Let us know if you need help with the tests.

**Comment by jameslamb** [comment]
@Yard1 since you're a new contributor, we'll have to manually approve CI runs on each commit

image

This is something GitHub added last year (I think?) to combat abuse of GitHub Actions.

Sorry for the inconvenience!

**Comment by Yard1** [comment]
@jameslamb All's good, was on the accepting side myself many times 😂 The tests pass locally for me now, so it should be good!

File: python-package/lightgbm/callback.py
Comment (jameslamb) [inline python-package/lightgbm/callback.py:360]: Could you please revert indentation changes like these? That would simplify the diff and make it easier for reviewers to understand what is being changed.

For example, on `master` this project has

```python
eval_result_str = '\t'.join([_format_eval_result(x) for x in best_score_list[i]])
```

and on this PR you have

```python
eval_result_str = '\t'.join(
                            [_format_eval_result(x) for x in self.best_score_list[i]])
```

We don't enforce a limit on line length in this project's Python code, and prefer to leave style-only changes to maintainers in most cases (to ensure that the git blame is easy to use as a source of information about when lines changed).

File: python-package/lightgbm/callback.py
Comment (jameslamb) [inline python-package/lightgbm/callback.py:236]:
```suggestion
        def __init__(
            self,
            stopping_rounds: int,
            first_metric_only: bool = False,
            verbose: bool = True,
            min_delta: Union[float, List[float]] = 0.0
        ) -> None:
```

Thanks for adding type hints here! Could you please use this suggested indentation? We've slowly been adopting type hints in this package (see the discussion and PRs linked to #3756), and have used the convention of this one-argument-per-line formatting when there is more than one argument.

For example:

https://github.com/microsoft/LightGBM/blob/717631af1836be9b03444ece485f3a3b5dbae5da/python-package/lightgbm/dask.py#L167-L177

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for this! The fact that tests are passing without needing any modification is a very good sign 🙌 

Please see my first round of suggestions. I think it would be easier to review this PR if some of the style changes you've propose were reverted.

File: python-package/lightgbm/callback.py
Comment (Yard1) [reply python-package/lightgbm/callback.py:360]: Hey @jameslamb, I couldn't find any information on what formatting style to use so I just ran autopep8. Let me try to make the formatting as close to original as possible.

**Comment by Yard1** [comment]
@jameslamb I've applied your feedback, please take a look. Thanks!

**Review by jmoralez — APPROVED** [review]
Thank you for this! LGTM

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much for doing this!

Could you please add a test in https://github.com/Yard1/LightGBM/blob/patch-1/tests/python_package_test/test_engine.py that ensures that this callback is pickleable? 

The primary purpose of this PR, as I understand it, is to make the early stopping callback pickleable so that it can be used by frameworks like Ray that are serializing the arguments to `lgb.train()` to send them around to different process. If I'm right about that, I think it would be valuable to explicitly test that so that future changes to LightGBM don't accidentally break that behavior.

Ideally this would work with `pickle`, `joblib`, and `cloudpickle`...could you try that? See the Dask tests for inspiration.

https://github.com/microsoft/LightGBM/blob/f185695617d29e446e021fed1261ba84f09ff992/tests/python_package_test/test_dask.py#L237-L260

https://github.com/microsoft/LightGBM/blob/f185695617d29e446e021fed1261ba84f09ff992/tests/python_package_test/test_dask.py#L1296

---------------------

@jmoralez I suspect we'll need to make changes similar to this PR to for all other callbacks to support `callbacks` in `lightgbm.dask`. I can write up a feature request for that.

**Comment by Yard1** [comment]
@jameslamb I have added a test, please take a look!

**Comment by Yard1** [comment]
The tests are failing due to a scheduled Windows brownout in CI.

**Comment by Yard1** [comment]
Hey @jameslamb the Windows CI env doesn't seem to have cloudpickle installed. How can we fix that?

File: python-package/lightgbm/callback.py
Comment (jameslamb) [inline python-package/lightgbm/callback.py:202]: can you please explain why it was necessary to pull this class out of the function? Was that required to make the callback function pickleable?

**Review by jameslamb — CHANGES_REQUESTED** [review]
Great test! Thanks very much for adding that, and for moving the pickling / unpickling helper functions into `utils.py`.

Please see one question I have about related changes in `callback.py`.

File: python-package/lightgbm/callback.py
Comment (Yard1) [reply python-package/lightgbm/callback.py:202]: Yes - cloudpickle can pickle local functions and classes, but pickle and joblib require them to be accessible from module level.

Comment (jameslamb) [reply python-package/lightgbm/callback.py:202]: Ah interesting, thanks for that. Glad we added tests!

**Comment by Yard1** [comment]
@jameslamb Had to merge with master, please rerun tests

File: tests/python_package_test/test_engine.py
Comment (Yard1) [inline tests/python_package_test/test_engine.py:23]:
```suggestion
from .utils import (load_boston, load_breast_cancer, load_digits, load_iris, make_synthetic_regression, pickle_obj,
                    unpickle_obj, sklearn_multiclass_custom_objective, softmax)
```

File: python-package/lightgbm/callback.py
Comment (StrikerRUS) [inline python-package/lightgbm/callback.py:352]:
```suggestion
    callback : _EarlyStoppingCallback
```

File: python-package/lightgbm/callback.py
Comment (StrikerRUS) [inline python-package/lightgbm/callback.py:358]: I think we don't need this line for the internal class.

File: python-package/lightgbm/callback.py
Comment (StrikerRUS) [inline python-package/lightgbm/callback.py:279]: I think we can encapsulate `_gt_delta()` and `_lt_delta()` functions in this class.

File: tests/python_package_test/test_engine.py
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:3430]: Please create a new file named `test_callback.py` in the `python_package_test` folder and move this test there.

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
Thank you very much for your work! Please consider checking some very minor suggestions below from me.

**Review by StrikerRUS — APPROVED** [review]
Awesome work! Thank you very much for this PR!

**Comment by Yard1** [comment]
@jameslamb Gentle ping :)

**Comment by Yard1** [comment]
Let me merge master again, maybe that will fix the CI

**Comment by Yard1** [comment]
@StrikerRUS @jameslamb CI is green now 🙏

**Comment by StrikerRUS** [comment]
@jameslamb Could you please take another look before merging?

**Review by jameslamb — APPROVED** [review]
Looks great!

Thanks so much for your work on this, and for adding tests. 🚀

**Comment by jameslamb** [comment]
Thanks again for this!

I changed the title to "make `early_stopping` callback pickleable" because PR titles in this project automatically become items in the release notes, and I think the fact that `lgb.early_stopping` is now pickleable will be more important to users than the particular mechanism used to achieve that.

I've also added #5080 documenting the work to ensure that all other callbacks are pickleable. Thanks for bringing the issue to our attention.

Good luck with `lightgbm-ray`, and we hope we'll see more contributions from you in the future 👋

**Comment by Yard1** [comment]
Thanks!

**Pull request merged**
